pkg/controller/istiocontrolplane/inputs.go

```diff
@@ -0,0 +1,91 @@
+package istiocontrolplane
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/helm/pkg/manifest"
+
+	"istio.io/operator/pkg/apis/istio/v1alpha1"
+	"istio.io/operator/pkg/helmreconciler"
+)
+
+// defaultProcessingOrder for the rendered charts
+var defaultProcessingOrder = []string{
+	"istio",
+	"istio/charts/security",
+	"istio/charts/galley",
+	"istio/charts/prometheus",
+	"istio/charts/mixer",
+	"istio/charts/pilot",
+	"istio/charts/gateways",
+	"istio/charts/sidecarInjectorWebhook",
+	"istio/charts/grafana",
+	"istio/charts/tracing",
+	"istio/charts/kiali",
+}
+
+// IstioInputFactory creates a RenderingInput specific to an IstioControlPlane instance.
+type IstioInputFactory struct{}
+
+var _ helmreconciler.RenderingInputFactory = &IstioInputFactory{}
+
+// IstioRenderingInput is a RenderingInput specific to an IstioControlPlane instance.
+type IstioRenderingInput struct {
+	instance *v1alpha1.IstioControlPlane
+}
+
+var _ helmreconciler.RenderingInput = &IstioRenderingInput{}
+
+// NewRenderingInput creates a new IstioRenderiongInput for the specified instance.
+func (f *IstioInputFactory) NewRenderingInput(instance runtime.Object) (helmreconciler.RenderingInput, error) {
+	istioControlPlane, ok := instance.(*v1alpha1.IstioControlPlane)
+	if !ok {
+		return nil, fmt.Errorf("object is not an IstioControlPlane resource")
+	}
+	return &IstioRenderingInput{instance: istioControlPlane}, nil
+}
+
+// GetChartPath returns the absolute path locating the charts to be rendered.
+func (i *IstioRenderingInput) GetChartPath() string {
+	path := i.instance.Spec.ChartPath
+	if len(path) == 0 {
+		return filepath.Join(controllerConfiguration.BaseChartPath, controllerConfiguration.DefaultChartPath)
+	} else if filepath.IsAbs(path) {
+		return i.instance.Spec.ChartPath
+	}
+	return filepath.Join(controllerConfiguration.BaseChartPath, path)
```

Comment:
Seems like it might be better to bind the path value when you create the renderer, or even the input to it. That way you don't need to do work every time to get the path. Also, IstioRenderingInput should probably not concern itself with logic about how the path is constructed.

Reply:
agreed, although it's only called once.

as for it being part of the rendering input, it is part of the input for rendering, so i would argue the logic belongs here.  that said, i could see logic being part of args.go, as that is what's driving the logic.  either way, i don't really care.  i'll move it to args.go.